File: code/ch12/12.2.3.PredictTheWinner.py
from typing import List


class Solution:
    def PredictTheWinner(self, nums: List[int]) -> bool:
        n = len(nums)
        if n % 2 == 0 or n == 1:
            return True
        dp = [0] * n

        for i in range(n - 1, -1, -1):
            dp[i] = nums[i]
            for j in range(i + 1, n):
                dp[j] = max(nums[i] - dp[j], nums[j] - dp[j - 1])
        return dp[-1] >= 0

File: code/ch12/12.2.3.PredictTheWinner.cpp
class Solution {
public:
    bool PredictTheWinner(vector<int>& nums) {
        int len = nums.size();
        
        if (len % 2 == 0 || len == 1) return true;
        
        vector<int> dp(len, 0);
        
        for (int i = len - 1; i > -1; i--) {
            dp[i] = nums[i];
            for (int j = i + 1; j < len; j++) {
                dp[j] = max(nums[i] - dp[j], nums[j] - dp[j - 1]);
            }
        }
        
        return dp[len - 1] >= 0; 
    }
};
